pkg/download/buffered_reader.go

```diff
@@ -35,57 +39,58 @@ func newBufferedReader(pool *bufferPool) *bufferedReader {
 // pool.
 func (b *bufferedReader) Read(buf []byte) (int, error) {
 	<-b.ready
+	if b.buf == nil {
+		return 0, io.EOF
+	}
 	n, err := b.buf.Read(buf)
 	// If we've read all the data,
-	if b.buf.Len() == 0 && b.buf != emptyBuffer {
+	if b.buf.Buffered() == 0 {
 		// return the buffer to the pool
 		b.pool.Put(b.buf)
 		// and replace our buffer with something that will always return EOF on
 		// future reads
-		b.buf = emptyBuffer
+		b.buf = nil
 	}
 	return n, err
 }
 
-func (b *bufferedReader) done() {
-	close(b.ready)
+func (b *bufferedReader) ReadFrom(r io.Reader) (int64, error) {
+	b.buf.Reset(r)
+	bytes, err := b.buf.Peek(b.buf.Size())
+	if err == io.EOF {
+		// ReadFrom does not return io.EOF
+		err = nil
+	}
```

Comment:
While this is generally unlikely, it is possible for the peek to not completely fill the buffer before 100 iterations. This could result in a contentLength mismatch as it would return fewer than requested bytes. This likely needs to be a for {} loop and handle the case of `io.ErrNoProgress`, another error (exit) or until the buffer's filled to expected content length

https://cs.opensource.google/go/go/+/refs/tags/go1.22.1:src/bufio/bufio.go;l=109

Reply:
if peek doesn't completely fill the buffer, we get another chance to fill the buffer when `Read` is called, because bufio.Reader is ultimately a wrapper around an underlying Reader. We are doing best effort to have the buffer full before the consumer calls Read, but if we aren't successful the behavior is still correct.

The downside is we effectively give up trying to download until the consumer gets around to reading this particular Reader. I think that's ok - this is "generally unlikely" as you say and I'm happy for it to not be 100% optimized.

Reply:
Yes, additionally this code in consistent_hashing.go:

```go
					n, err := br.ReadFrom(resp.Body)
					if err != nil {
						return err
					} else if n != contentLength {
						return ErrContentLengthMismatch{downloadedBytes: n, contentLength: contentLength}
					}
```

Will not let us get to the Read during the consumer, and we still need to use the .Done() call for the channel before we read. As implemented if we hit the unlikely scenario we might deadlock the download?

Reply:
the call do br.Done() is in a defer at the top of that block so I'm unclear what problem you're describing here?

Reply:
I pushed a commit to do a loop checking for progress.